mono/metadata/threads.c

```diff
@@ -5065,15 +5088,67 @@ threads_add_joinable_thread_nolock (gpointer tid)
 }
 #endif
 
+static void
+threads_add_pending_joinable_runtime_thread (gpointer thread_info)
+{
+	g_assert (thread_info);
+	MonoThreadInfo *mono_thread_info = (MonoThreadInfo*)thread_info;
+
+	if (mono_thread_info->runtime_thread) {
+		gint32 old_joinable_state = mono_atomic_load_i32 (&mono_thread_info->thread_joinable_state);
+		if (old_joinable_state == THREAD_JOINABLE_STATE_NONE) {
+			if (mono_atomic_cas_i32 (&mono_thread_info->thread_joinable_state, THREAD_JOINABLE_STATE_PENDING, THREAD_JOINABLE_STATE_NONE) == THREAD_JOINABLE_STATE_NONE) {
+				gint32 result = mono_atomic_inc_i32 (&pending_joinable_runtime_thread_count);
+				g_assert (result > 0);
+			}
+		}
+	}
+}
+
+static gboolean
+threads_wait_pending_joinable_runtime_threads (uint32_t timeout)
+{
+	gboolean wait_result = TRUE;
+	if (mono_atomic_load_i32 (&pending_joinable_runtime_thread_count) > 0) {
+		joinable_threads_lock ();
+		if (timeout == MONO_INFINITE_WAIT) {
```

Comment:
Doesn't mono_os_cond_timedwait handle MONO_INFINITE_WAIT correctly?

The infinite timeout case doesn't handle spurious wakeups and bail out without checking the result.

Reply:
There are multiple problems with the mono_os_cond_timedwait that triggered this implementation.

* There is no way give it an ability to recheck to handle spurious wakeups meaning that the caller needs to handle this.
* No way to make a difference between a timeout and an error.

The above implementation follow the implementation from mono_os_event_wait_multiple, it has the same issue with infinite timeout, not doing the retry on spurious wakeups, 64-bit, 32-bit, singed unsinged etc.

I can of course adjust the implementation above to handle spurious wakeups, but I believe the better alternative is to add a new mono_os_cond_timedwait that handles spursious wakeups using a recheck function and that it will either succeed or timeout. Thoughts?